### PR TITLE
Improve Mean Average Precision performance

### LINKED ISSUE #1024: MeanAveragePrecision is slow
## 🐛 Bug

It's extremely slow to compute the mean-average-precision since torchmetrics > 0.6.0.

### To Reproduce

I noticed that my training times have almost doubled since I upgraded torchmetrics from 0.6.0, because validation using the MAP / MeanAveragePrecision metric is so much slower. During validation steps I call `update()`, and in the end of a validation epoch I call `compute()` on the `MeanAveragePrecision` object.

I calculated the time that spent inside `compute()` with different torchmetrics versions:

- torchmetrics 0.6.0: 12 s
- torchmetrics 0.6.1: didn't work for some reason
- torchmetrics 0.6.2: 9.5 min
- torchmetrics 0.7.0: 9.4 min
- torchmetrics 0.7.1: 1.9 min
- torchmetrics 0.7.2: 2.0 min
- torchmetrics 0.7.3: 1.9 min
- torchmetrics 0.8.0: 4.5 min
- torchmetrics 0.8.1: 4.6 min
- torchmetrics 0.8.2: 4.6 min

It seems that after 0.6.0 the time to run `compute()` has increased from 10 seconds to 9.5 minutes. In 0.7.1 it was improved and took 2 minutes. Then in 0.8.0 things got worse again and it took 4.5 minutes to run `compute()`. This is more than 20x slower than with 0.6.0 and for example when training 100 epochs adds another 7 hours to the training time.

### Environment

- TorchMetrics version (and how you installed TM, e.g. `conda`, `pip`, build from source): 0.6.0 through 0.8.2, installed using pip
- Python & PyTorch Version (e.g., 1.0): Python 3.8.11, PyTorch 1.10.0
- Any other relevant information such as OS (e.g., Linux): Linux

### PR DESCRIPTION
**Pull request by wilderrodrigues**
## What does this PR do?

Fixes #1024

**Update:** The initial PR contained erroneous information about the running time due to a silent issues that was happening, killing the run because some tensors were in a different device: CPU instead of GPU and vice-versa. The changes are still interesting as the performance increased a lot.

It seems that we faced some regression between version 0.6.0 and the latest releases. With the refactor of moving away from pycocotools and the addition of new operations, the appropriate device was not being used. That behaviour lead to unnecessary transfer of computation between GPU and CPU.

Based on the code provided in the issue #1024 discussion, I also added a couple of unit tests to cover for computation with GPUs and CPUs. Since the results might vary a bit, depending on the GPU/CPU model, I added a relative tolerance of 1.0 trying to cover for the gap.

Some results:

Before (GPU):
```
Total time in init: 1.0306465921457857
Total time in update: 0.0780688391532749
Total time in compute: 241.5502170859836
```

After (GPU):
```
Total time in init: 1.1004462020937353
Total time in update: 0.08146677981130779
Total time in compute: 137.3318270179443
```

At it has been pointed out, those computations when performed on the CPU might be faster. The important part is making sure that the device is properly allocated to the tensors. When running the same tests on CPU I got the following results:

```
Total time in init: 0.020693536149337888
Total time in update: 0.08699228917248547
Total time in compute: 72.37102157599293
```

I will run some tests with our production model and share the deltas here.

## Before submitting

- [X] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [X] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Yep! This was fun. After trying different versions and commit hashes, I just decided to clone, fix and push the PR. It will help us a lot and I'm sure it will help out the community.

**Test with real data**

- 2 RTX 3090
- VisDrone dataset

image

The pink line is the result with the same run using the master branch (which is faster than the latest 0.11 release)
Don't bother about the flat line at the end of each run. That's just PyCharm holding the GPU busy before I clicked on stop.

### CODE REVIEW

**Comment by Borda** [comment]
wow, very nice boost!

**Comment by wilderrodrigues** [comment]
> wow, very nice boost!

I marked it as WIP for now because there is sill something wrong with it. The initial results could be wrong due to some internal error that was not picked up. For instance, some tensors being on the CPU and other on the GPU. I’m looking into it right now.

**Comment by wilderrodrigues** [comment]
My apologies, @Borda. It seems that when I rebased with master and force-pushed, a lot of other things came up to the PR. It looks a bit overwhelming now to review. Could have also been the pre-commit I ran.

**Comment by Borda** [comment]
@wilderrodrigues no problem :)

File: src/torchmetrics/detection/mean_ap.py [hunk 1]
```diff
@@ -447,7 +450,7 @@ def _get_safe_item_values(self, item: Dict[str, Any]) -> Union[Tensor, Tuple]:
         elif self.iou_type == "segm":
             masks = []
 
-            for i in item["masks"].cpu().numpy():
+            for i in item["masks"].to(self.device).numpy():
```
[the remark this answers is not in the document]
Comment (wilderrodrigues) [reply src/torchmetrics/detection/mean_ap.py:453]: Good catch, no longer needed.

File: tests/unittests/detection/benchmark/benchmark_setup.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import time
```
[the remark this answers is not in the document]
Comment (wilderrodrigues) [reply tests/unittests/detection/benchmark/benchmark_setup.py:1]: Will move it there.

[the remark this answers is not in the document]
Comment (wilderrodrigues) [reply tests/unittests/detection/benchmark/benchmark_setup.py]: I moved the `benchmark` packaged under the `unittests` one.

File: src/torchmetrics/detection/mean_ap.py [hunk 3]
```diff
@@ -601,93 +603,80 @@ def _evaluate_image(
...
+            "gtIgnore": gt_ignore,
+            "dtIgnore": det_ignore,
         }
 
     @staticmethod
-    def _find_best_gt_match(
-        thr: int, gt_matches: Tensor, idx_iou: float, gt_ignore: Tensor, ious: Tensor, idx_det: int
-    ) -> int:
-        """Return id of best ground truth match with current detection.
+    def _find_best_gt_matches(thr: Tensor, gt_matches: Tensor, gt_ignore: Tensor, ious: Tensor) -> Tensor:
```
[the remark this answers is not in the document]
Comment (wilderrodrigues) [reply src/torchmetrics/detection/mean_ap.py:661]: Looking into this one. The call to the function is now vectorised and takes into account the batch.

**Comment by wilderrodrigues** [comment]
Just a quick update on this PR:

1. We have consistency across multiple runs w.r.t runtime (see fig. 1)
2. We have consistency - determinism - on our metrics across multiple runs
    - See fig. 2 for Val mAP and fig.3 for Test mAP

- 2 RTX 3090 GPUs
- VisDrone dataset
- 32 batch size
- 16 workers
- batch limits for train, val and test is 50
- 10 epochs each run, validating every 2 epochs

image

image

image

**Comment by wilderrodrigues** [comment]
Can this one get some love, @Borda @SkafteNicki @justusschock @tchaton :)

File: src/torchmetrics/detection/mean_ap.py [hunk 4]
```diff
@@ -475,11 +478,11 @@ def _compute_iou(self, idx: int, class_id: int, max_det: int) -> Tensor:
         """
 
         # if self.iou_type == "bbox":
-        gt = self.groundtruths[idx]
-        det = self.detections[idx]
+        gt = self.groundtruths[idx].to(self.device)
```
Comment (Borda) [inline src/torchmetrics/detection/mean_ap.py:481]: these are tuples so you can't move them simply to the device, you need to do it per item or can we keep them in CPU?

Comment (wilderrodrigues) [reply src/torchmetrics/detection/mean_ap.py:481]: They are not tuples, the are list of tensors. So, by accessing an element via the index, we get a Tensor that can be moved to the device.

And I can see it through a run:

```
type(self.groundtruths[idx])
<class 'torch.Tensor'>
```

And the list is declared on line 360:

```
groundtruths: List[Tensor]
```

And the same for the other one:

```
type(self.detections[idx])
<class 'torch.Tensor'>
```

@Borda

Comment (Borda) [reply src/torchmetrics/detection/mean_ap.py:481]: ok, so there are some data types which are tuple as you can see from the failing tests :)

Comment (wilderrodrigues) [reply src/torchmetrics/detection/mean_ap.py:481]: Ah. Those tests were passing before I rebased with the Master branch. But let me check them and understand what's going on.

Comment (wilderrodrigues) [reply src/torchmetrics/detection/mean_ap.py:481]: Hey @Borda ,

All failing jobs relate to this:

```
INFO:root:requirements_path='requirements.txt' with torch_version='None'
INFO:root:finding ecosystem versions for: 1.13.1
Traceback (most recent call last):
  File "./requirements/adjust-versions.py", line 83, in <module>
    requirements = adjust(requirements, torch_version)
  File "./requirements/adjust-versions.py", line 64, in adjust
    latest = find_latest(torch_version)
  File "./requirements/adjust-versions.py", line 51, in find_latest
    raise ValueError(f"Missing {ver} in {VERSIONS}")
ValueError: Missing 1.13.1 in [{'torch': '1.13.0', 'torchvision': '0.14.0', 'torchtext': '0.14.0'}, {'torch': '1.12.1', 'torchvision': '0.13.1', 'torchtext': '0.13.1'}, {'torch': '1.12.0', 'torchvision': '0.13.0', 'torchtext': '0.13.0'}, {'torch': '1.11.0', 'torchvision': '0.12.0', 'torchtext': '0.12.0'}, {'torch': '1.10.2', 'torchvision': '0.11.3', 'torchtext': '0.11.2'}, {'torch': '1.10.1', 'torchvision': '0.11.2', 'torchtext': '0.11.1'}, {'torch': '1.10.0', 'torchvision': '0.11.1', 'torchtext': '0.11.0'}, {'torch': '1.9.1', 'torchvision': '0.10.1', 'torchtext': '0.10.1'}, {'torch': '1.9.0', 'torchvision': '0.10.0', 'torchtext': '0.10.0'}, {'torch': '1.8.2', 'torchvision': '0.9.1', 'torchtext': '0.9.1'}, {'torch': '1.8.1', 'torchvision': '0.9.1', 'torchtext': '0.9.1'}, {'torch': '1.8.0', 'torchvision': '0.9.0', 'torchtext': '0.9.0'}]
```

Any idea what's going on? I mean, it's clear is version related, but why it changed after the rebase?

Comment (justusschock) [reply src/torchmetrics/detection/mean_ap.py:481]: @wilderrodrigues this is just bad luck as torch 1.13.1 was released between the test runs.

**Review by justusschock — APPROVED** [review]

**Comment by justusschock** [comment]
@wilderrodrigues sorry for the long return time in this review. Your PR looks good to me!

File: src/torchmetrics/detection/mean_ap.py (hunk 4 above)
Comment (Borda) [reply src/torchmetrics/detection/mean_ap.py:481]: > All failing jobs relate to this:

it was fixed on master

**Comment by wilderrodrigues** [comment]
> @wilderrodrigues sorry for the long return time in this review. Your PR looks good to me!

Thanks, @Borda .

Ben also a bit away in the last few days, a lot of work to catch up with before the holidays season. I will rebase the PR and wait for the merge to happen. :)

**Comment by wilderrodrigues** [comment]
A second approval and we can merge this one, @Borda .

**Comment by Borda** [comment]
> A second approval and we can merge this one, @Borda .

could we pls fist fix the failing tests :)

**Comment by wilderrodrigues** [comment]
> > A second approval and we can merge this one, @Borda .
> 
> could we pls first fix the failing tests :)

Oh yeah, sure. I haven't seen them.

**Comment by wilderrodrigues** [comment]
Most of the tests have been fixe. I'm now trying to fix the `test_map_box` as it is showing a discrepancy in the mAP results.

The assertion is not passing during the CI run:

```
pl_result = {'map': tensor(0.6477)

sk_result = array([0.637], dtype=float32), atol = 0.01, key = 'map')
```

As we can see, 0.0107 difference.

What is getting me intrigued is that running on my RTX-3090 is passing just fine (see below).

```
(torchmetrics) wilderr@XXXXXXXXX:/mnt/vidint_ssd/wilder/remote_metrics$ pytest tests/unittests/detection/test_map.py::TestMAP::test_map_bbox
============================================================================================ test session starts ============================================================================================
platform linux -- Python 3.9.0, pytest-7.2.0, pluggy-1.0.0
rootdir: /mnt/vidint_ssd/wilder/remote_metrics, configfile: setup.cfg
collected 2 items                                                                                                                                                                                                                  

tests/unittests/detection/test_map.py ..                                                                                                                                                                                     [100%]

============================================================================================ 2 passed, 2 warnings in 31.49s ============================================================================================
(torchmetrics) wilderr@XXXXXXXXX:/mnt/vidint_ssd/wilder/remote_metrics$ 

```

**Comment by wilderrodrigues** [comment]
> 

It fails with on those cases:

```
unittests/detection/test_map.py::TestMAP::test_map_bbox[False-True]
unittests/detection/test_map.py::TestMAP::test_map_bbox[True
```

However, those are passing. It makes life hard to debug if they work on my MacBook and RTX-3090.

image

image

On the CI environment, the issue is intermittent: in some environment, the test fails twice, with 2 different parameters sets; and some others it fails just once. The gap is always the same: 0.0007 above the tolerance.

Need some ideas here, @Borda

**Comment by senarvi** [comment]
I tried this in a training job and got this error:

```
Traceback (most recent call last):                                                                                                                                                                                                                                                        
  ...
  File ".../torchmetrics/metric.py", line 542, in wrapped_func
    value = compute(*args, **kwargs)
  File ".../torchmetrics/detection/mean_ap.py", line 888, in compute
    precisions, recalls = self._calculate(classes)
  File ".../torchmetrics/detection/mean_ap.py", line 762, in _calculate
    recall, precision, scores = MeanAveragePrecision.__calculate_recall_precision_scores(
  File ".../torchmetrics/detection/mean_ap.py", line 861, in __calculate_recall_precision_scores
    pr[pr.isnan()] = torch.tensor(0).to(recall.device)
RuntimeError: Index put requires the source and destination dtypes match, got Float for the destination and Long for the source.
```

**Comment by Borda** [comment]
> Index put requires the source and destination dtypes match, got Float for the destination and Long for the source.

I think we are fine to retype it for the test comparison so both will be float in this case :otter:

**Comment by wilderrodrigues** [comment]
> I tried this in a training job and got this error:
> 
> ```
> Traceback (most recent call last):                                                                                                                                                                                                                                                        
>   ...
>   File ".../torchmetrics/metric.py", line 542, in wrapped_func
>     value = compute(*args, **kwargs)
>   File ".../torchmetrics/detection/mean_ap.py", line 888, in compute
>     precisions, recalls = self._calculate(classes)
>   File ".../torchmetrics/detection/mean_ap.py", line 762, in _calculate
>     recall, precision, scores = MeanAveragePrecision.__calculate_recall_precision_scores(
>   File ".../torchmetrics/detection/mean_ap.py", line 861, in __calculate_recall_precision_scores
>     pr[pr.isnan()] = torch.tensor(0).to(recall.device)
> RuntimeError: Index put requires the source and destination dtypes match, got Float for the destination and Long for the source.
> ```

Hey @senarvi , thanks for reporting this. Fixing it here.

It's a bit weird it was not caught in in the tests.

**Comment by wilderrodrigues** [comment]
> 

This `torch.tensor(0)` should be like this `torch.tensor(0.)`

Pushing a fix and you can go ahead again.

**Review by SkafteNicki — APPROVED** [review]

**Comment by justusschock** [comment]
@wilderrodrigues could you check why the results are not matching?

**Comment by wilderrodrigues** [comment]
> @wilderrodrigues could you check why the results are not matching?

Hi @Borda,

Sorry for the delay in my responses: just back from some days off and get through everything.

Yeah, I will put some time on it and try to get those results to match. It's really weird to have it working on 2 different machines and not on CI. I mean, it's not the usual "it works on my computer". haha A clean checkout of the branch and installation on a RTX 3090 works out of the box, just fine. Will investigate today/tomorrow so we can get this one merged.

**Comment by senarvi** [comment]
I tested this pull request on real-world data and it's a considerable improvement in terms of speed.

- The current master: around 350 seconds per epoch
- This pull request: around 100 seconds per epoch
- The old version that uses pycocotools: around 15 seconds per epoch

Unfortunately my experiment also seems to indicate some problem with the code. With the current master version and the pycocotools version I get identical results, but using the code from this PR I get higher scores.

**Comment by wilderrodrigues** [comment]
> I tested this pull request on real-world data and it's a considerable improvement in terms of speed.
> 
> * The current master: around 350 seconds per epoch
> * This pull request: around 100 seconds per epoch
> * The old version that uses pycocotools: around 15 seconds per epoch
> 
> Unfortunately my experiment also seems to indicate some problem with the code. With the current master version and the pycocotools version I get identical results, but using the code from this PR I get higher scores.

Hi @senarvi ,

Do you have a benchmark of the results? I would like to see the difference in the scores. I'm trying to make some time tomorrow to dig into it and get it fixed for good.

**Comment by senarvi** [comment]
> Do you have a benchmark of the results? I would like to see the difference in the scores. I'm trying to make some time tomorrow to dig into it and get it fixed for good.

Great! Unfortunately I cannot share this benchmark, as it's proprietary data.

**Comment by wilderrodrigues** [comment]
> > Do you have a benchmark of the results? I would like to see the difference in the scores. I'm trying to make some time tomorrow to dig into it and get it fixed for good.
> 
> Great! Unfortunately I cannot share this benchmark, as it's proprietary data.

Okay, @senarvi , I totally understand.

But could you at least tell me in which range the gap is? For instance, AP 0.7 with pycocotools and with these changes the AP goes to 0.85. Just for me to have an idea.

Thanks!

**Comment by senarvi** [comment]
The difference was considerable. There's a [benchmark in this issue](https://github.com/Lightning-AI/metrics/issues/1024#issuecomment-1281424230) that uses random data. Using that script I get the following difference:

pycocotools version: `{'map': tensor(0.1534), 'map_50': tensor(0.5260), 'map_75': tensor(0.0336), 'map_small': tensor(0.1534), 'map_medium': tensor(-1.), 'map_large': tensor(-1.), 'mar_1': tensor(0.0449), 'mar_10': tensor(0.3039), 'mar_100': tensor(0.5445), 'mar_small': tensor(0.5445), 'mar_medium': tensor(-1.), 'mar_large': tensor(-1.), 'map_per_class': tensor(-1.), 'mar_100_per_class': tensor(-1.)}`

this PR: `{'map': tensor(0.2222), 'map_50': tensor(0.7135), 'map_75': tensor(0.0594), 'map_small': tensor(0.2222), 'map_medium': tensor(-1.), 'map_large': tensor(-1.), 'mar_1': tensor(0.0449), 'mar_10': tensor(0.4453), 'mar_100': tensor(2.2028), 'mar_small': tensor(2.2028), 'mar_medium': tensor(-1.), 'mar_large': tensor(-1.), 'map_per_class': tensor(-1.), 'mar_100_per_class': tensor(-1.), 'classes': tensor([0, 1], dtype=torch.int32)}`

Maybe you can debug it by generating random data and observing what kind of data causes a difference.

**Comment by lucasc896** [comment]
These performance gains look great! Are there any plans to merge this soon?

**Comment by wilderrodrigues** [comment]
> These performance gains look great! Are there any plans to merge this soon?

Hi @lucasc896 ,

I have to benchmark the changes against pycocotools as some discrepancies have been mentioned.

I did benchmark against the previous version of Torchmetrics before the MR was created, and it was fine. Have to see what's going on.

Most of the unit tests are fine and they also cover for the benchmark. The only thing is that 1 case have a mAP of 0.0007 above the expected threshold and then fails. Will make time to look into it this week.

**Review by Borda — APPROVED** [review]

**Comment by PabloMessina** [comment]
Hello guys, I came across this pull request precisely because I'm facing performance issues with torchmetrics MAP right now.

This is my code:

```python
import torch
from torchmetrics.detection.mean_ap import MeanAveragePrecision
import random

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

def _generate_random_bbox():
    xmin =  random.random()
    ymin =  random.random()
    xmax = xmin + random.random() * (1 - xmin)
    ymax = ymin + random.random() * (1 - ymin)
    return [xmin, ymin, xmax, ymax]

N_examples = 3000
N_classes = 36

preds = [
  dict(
    boxes=torch.tensor([_generate_random_bbox() for _ in range(N_classes)]).to(device),
    scores=torch.tensor([random.random() for _ in range(N_classes)]).to(device),
    labels=torch.tensor([i for i in range(N_classes)]).to(device),
  )
    for _ in range(N_examples)
]

target = [
  dict(
    boxes=torch.tensor([_generate_random_bbox() for _ in range(N_classes - 4)]).to(device),
    labels=torch.tensor([i for i in random.sample(range(N_classes), N_classes - 4)]).to(device),
  )
    for _ in range(N_examples)
]

import time
start = time.time()
metric = MeanAveragePrecision(
    iou_thresholds=[0.1, 0.5, 0.9],
    rec_thresholds=[0.0, 0.3, 0.7, 1.0],
    max_detection_thresholds=[N_classes],
)
metric.update(preds, target)
from pprint import pprint
output = metric.compute()
pprint(output)
print('Elased time:', time.time() - start)
```

And this is the output I'm getting:

>  {'map': tensor(-1.),
 'map_50': tensor(0.0039),
 'map_75': tensor(-1),
 'map_large': tensor(-1.),
 'map_medium': tensor(-1.),
 'map_per_class': tensor(-1.),
 'map_small': tensor(0.0186),
 'mar_36': tensor(0.0315),
 'mar_36_per_class': tensor(-1.),
 'mar_large': tensor(-1.),
 'mar_medium': tensor(-1.),
 'mar_small': tensor(0.0315)}
Elased time: 340.79711151123047

So my main concern is that this is taking too much time to compute. A batch of 36 classes and 3000 examples, which is a reasonable estimate of the sizes of my validation sets, is taking 340 seconds (5 minutes and 40 seconds), which is a lot. Another concern I have is the fact that I'm getting -1.0 as map. Why is it so? How should I interpret -1.0 as output? I thought that map ranges between 0 and 1, so I guess I'm probably doing something wrong with my implementation above.

Thanks in advance.

**Comment by wilderrodrigues** [comment]
> Hello guys, I came across this pull request as I was searching for solutions to the performance issues with torchmetrics MAP I'm having right now.
> 
> This is my code:
> 
> ```python
> import torch
> from torchmetrics.detection.mean_ap import MeanAveragePrecision
> import random
> 
> device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
> 
> def _generate_random_bbox():
>     xmin =  random.random()
>     ymin =  random.random()
>     xmax = xmin + random.random() * (1 - xmin)
>     ymax = ymin + random.random() * (1 - ymin)
>     return [xmin, ymin, xmax, ymax]
> 
> N_examples = 3000
> N_classes = 36
> 
> preds = [
>   dict(
>     boxes=torch.tensor([_generate_random_bbox() for _ in range(N_classes)]).to(device),
>     scores=torch.tensor([random.random() for _ in range(N_classes)]).to(device),
>     labels=torch.tensor([i for i in range(N_classes)]).to(device),
>   )
>     for _ in range(N_examples)
> ]
> 
> target = [
>   dict(
>     boxes=torch.tensor([_generate_random_bbox() for _ in range(N_classes - 4)]).to(device),
>     labels=torch.tensor([i for i in random.sample(range(N_classes), N_classes - 4)]).to(device),
>   )
>     for _ in range(N_examples)
> ]
> 
> import time
> start = time.time()
> metric = MeanAveragePrecision(
>     iou_thresholds=[0.1, 0.5, 0.9],
>     rec_thresholds=[0.0, 0.3, 0.7, 1.0],
>     max_detection_thresholds=[N_classes],
> )
> metric.update(preds, target)
> from pprint import pprint
> output = metric.compute()
> pprint(output)
> print('Elapsed time:', time.time() - start)
> ```
> 
> And this is the output I'm getting:
> 
> > {'map': tensor(-1.),
> > 'map_50': tensor(0.0039),
> > 'map_75': tensor(-1),
> > 'map_large': tensor(-1.),
> > 'map_medium': tensor(-1.),
> > 'map_per_class': tensor(-1.),
> > 'map_small': tensor(0.0186),
> > 'mar_36': tensor(0.0315),
> > 'mar_36_per_class': tensor(-1.),
> > 'mar_large': tensor(-1.),
> > 'mar_medium': tensor(-1.),
> > 'mar_small': tensor(0.0315)}
> > Elapsed time: 340.79711151123047
> 
> So my main concern is that this is taking too much time to compute. A batch of 36 classes and 3000 examples, which is a reasonable estimate of the sizes of my validation sets, is taking 340 seconds (5 minutes and 40 seconds), which is a lot. Another concern I have is the fact that I'm getting -1.0 as map. Why is it so? How should I interpret -1.0 as output? I thought that map ranges between 0 and 1, so I guess I'm probably doing something wrong with my implementation above.
> 
> Thanks in advance.

Hi @PabloMessina ,

Thanks for taking time to have a look at it and also the PR. The code implemented here makes the metrics run in 1/3 of the time. I'm looking into what's left today and hopefully this will get merged this week.

Cheers!

File: src/torchmetrics/detection/mean_ap.py [hunk 5]
```diff
@@ -473,24 +476,19 @@ def _compute_iou(self, idx: int, class_id: int, max_det: int) -> Tensor:
...
         gt = [gt[i] for i in gt_label_mask]
         det = [det[i] for i in det_label_mask]
 
-        if len(gt) == 0 or len(det) == 0:
-            return Tensor([])
-
         # Sort by scores and use only max detections
-        scores = self.detection_scores[idx]
+        scores = self.detection_scores[idx].to(self.device)
         scores_filtered = scores[self.detection_labels[idx] == class_id]
```
[the remark this answers is not in the document]
Comment (Borda) [reply src/torchmetrics/detection/mean_ap.py:488]: Could you add it here as siggestion so we can eventually just accept it? :)

[the remark this answers is not in the document]
Comment (wilderrodrigues) [reply src/torchmetrics/detection/mean_ap.py:487]: I just rebased the PR with the latest master changes and I might have found the issues with the difference in results.

I will also push this change suggested by @ksachdeva

**Comment by wilderrodrigues** [comment]
@Borda , I'm now able to reproduce it locally. Will look into fixing this once and for all.

```
result["map"]
tensor(0.6477)
sk_result["map"]
tensor([0.6370])
```

**Comment by Borda** [comment]
> I'm now able to reproduce it locally. Will look into fixing this once and for all.

@SkafteNicki @Lightning-AI/core-metrics thoughts?

**Comment by wilderrodrigues** [comment]
New fixes coming in, @Borda @senarvi ... finally.

image

All tests are good now, just trying to improve speed a little bit more, as we now are above the previous (yet broken) implementation.

* Performance on CPU

```
Total time in init: 0.0009379680000023427
Total time in update: 0.11780578900001615
Total time in compute: 113.22202207199999
```

Current Master performance on CPU is about 160 seconds. The broken implementation was 77 seconds... but well, it was broken.

**Comment by ddelange** [comment]
Hi @wilderrodrigues 👋 

Out of curiosity: any chance you could run a substantial map calculation through [py-spy](https://pypi.org/project/py-spy/) and post the flamegraph here?

```
pip install py-spy
py-spy record -o profile.svg -- python myprogram.py
```

**Comment by senarvi** [comment]
> All tests are good now, just trying to improve speed a little bit more, as we now are above the previous (yet broken) implementation.

That's really great!

**Comment by wilderrodrigues** [comment]
> Hi @wilderrodrigues 👋
> 
> Out of curiosity: any chance you could run a substantial map calculation through [py-spy](https://pypi.org/project/py-spy/) and post the flamegraph here?
> 
> ```
> pip install py-spy
> py-spy record -o profile.svg -- python myprogram.py
> ```

Hi @ddelange ,

Sure. I'm busy with some last things in terms of speed and will run it once I push the last commits.

**Comment by Borda** [comment]
We can sequence it out to another PR, let's just get this green...

**Comment by Borda** [comment]
@wilderrodrigues, we value your work and thank you for your contribution :purple_heart: could you pls revisit this PR with respect to the recently merged revert to COCO implementation #1327 and setting the torch version as an Internal option? :rabbit:

**Comment by Borda** [comment]
Once again thank you for all your effort! :purple_heart: 
 Feel free to reopen this PR, and happy to talk about it any time in the future :chipmunk:

**Pull request closed without merging**